pkg/services/signingkeys/signingkeystore/store.go

```diff
@@ -0,0 +1,194 @@
+	privateKeyPEM, err := s.encodePrivateKey(ctx, privateKey)
+	key := &SigningKey{
+		KeyID:      keyID,
+		PrivateKey: privateKeyPEM,
+		AddedAt:    time.Now(),
+		Alg:        alg,
+		ExpiresAt:  expiresAt,
+	}
+
+	dbSession := s.dbStore.GetSqlxSession()
+	err = dbSession.WithTransaction(ctx, func(tx *session.SessionTx) error {
+		existingKey := SigningKey{}
+		err := tx.Get(ctx, &existingKey, "SELECT * FROM signing_key WHERE key_id = ?", keyID)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return err
+		}
+
+		if len(existingKey.PrivateKey) == 0 {
+			_, err = tx.Exec(ctx,
+				"INSERT INTO signing_key (key_id, private_key, added_at, alg, expires_at) VALUES (?, ?, ?, ?, ?)",
+				key.KeyID, key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt)
+			return err
+		}
+
+		if force || (existingKey.ExpiresAt != nil && existingKey.ExpiresAt.Before(time.Now())) {
+			_, err = tx.Exec(ctx,
+				"UPDATE signing_key SET private_key = ?, added_at = ?, alg = ?, expires_at = ? WHERE key_id = ?",
+				key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt, key.KeyID)
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetPrivateKey returns the private key with the specified key ID
+func (s *Store) GetPrivateKey(ctx context.Context, keyID string) (crypto.Signer, error) {
+	key := &SigningKey{}
+	err := s.dbStore.GetSqlxSession().Get(ctx, key, "SELECT * FROM signing_key WHERE key_id = ?", keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	signKey, err := s.decodePrivateKey(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return signKey, nil
+}
+
+func (s *Store) encodePrivateKey(ctx context.Context, privateKey crypto.Signer) ([]byte, error) {
+	// Encode private key to binary format
+	pKeyBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encode private key to PEM format
+	privateKeyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: pKeyBytes,
+	})
+
+	encrypted, err := s.secretsService.Encrypt(ctx, privateKeyPEM, secrets.WithoutScope())
+	if err != nil {
+		return nil, err
+	}
+
+	encoded := make([]byte, base64.StdEncoding.EncodedLen(len(encrypted)))
+	base64.StdEncoding.Encode(encoded, encrypted)
+	return encoded, nil
+}
+
+func (s *Store) decodePrivateKey(ctx context.Context, signingKey *SigningKey) (crypto.Signer, error) {
+	// Bail out if empty string since it'll cause a segfault in Decrypt
+	if len(signingKey.PrivateKey) == 0 {
+		return nil, errors.New("private key is empty")
+	}
+
+	payload := make([]byte, base64.StdEncoding.DecodedLen(len(signingKey.PrivateKey)))
+	_, err := base64.StdEncoding.Decode(payload, signingKey.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	decrypted, err := s.secretsService.Decrypt(ctx, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	block, _ := pem.Decode(decrypted)
+	if block == nil {
+		return nil, errors.New("failed to decode private key PEM")
+	}
+
```

Comment:
Just to be more defensive I think a check for block type would make sense, wdyt?
```
if block.Type != "PRIVATE KEY" {
	return nil, errors.New("invalid block type")
}
```

Reply:
+1